### ISSUE TITLE
PostgreSQL - Use adminer instead of phppgadmin?

### ISSUE THREAD

**Issue by DocCyblade** [turnkeylinux/tracker#386]
Since I have seen a push away from phpmyadmin should we convert away from phppgadmin?

TODO: 
~~1. Remove all trace of phpPgAdmin from build files~~
~~2. Add adminer to build files~~
~~3. Customize adminer install target PostgreSQL~~
~~4. Add Web Basic Web control panel at 80/443~~
~~5. Replace documentation verbiage to adminer from phpPgAdmin~~

Working branch https://github.com/DocCyblade/postgresql/tree/upgrade-to-adminer for those who want to track the code process

**Comment by l-arnold** [comment]
My understqnding was yes on Adminer.  Reference it if you can.

There are a few outside default used in some apps.  Look To Compare How Adminer Was Called In LAPP and Postgrea.

this folder may also hold some value in looking at and seeing how it gets called.
https://github.com/turnkeylinux/common/tree/master/conf

See the changes in zoster esquire and LAPP and can likely mimicking those.


Landis Arnold

Cell 303.601.0622
Office 303.444.2336

on mobile 
(so typos expected and excusable perhaps)

Sent from Nine

**Comment by DocCyblade** [comment]
Thats what I thought, already looking at lamp :-)

**Comment by DocCyblade** [comment]
Looks like postgres application is using lighttpd to power phpadmin, guess I should use lighttpd to power adminer as well. Can't user lapp as lapp uses apache.

**Comment by JedMeister** [comment]
TBH I was almost half thinking of leaving phpPgAdmin in the PostgreSQL appliance. Just because Adminer seems a little light on features with regard to administering PostgreSQL. From what I gather phpPgAdmin is a more powerful solution.

So leaving phpPgAdmin in the standalone DB appliance seemed like it mya be a good idea? 

Having said all that I haven't used PgSQL much myself so I would be really interested to hear from any of you guys who use it what you think... 

Does Adminer provide enough for PgSQL DB management? I think that's it's fine for an appliance that is primarily a web host with a PgSQL backend - but for a standalone DB server?

**Comment by l-arnold** [comment]
I vote fot either phpPgAdmin or possibly both.  I think both is ok as some will learn one or the other.

Same perhaps for mysql.  Does it also use lighttpd?

Was really awkward having Postgresql use lighttpd in terms of trying to implement SSL on Odoo but thats another subject.

**Comment by JedMeister** [comment]
So are you suggesting phpPgAdmin via the web root (i.e. https://example.com/) and Adminer via port 12322 (e.g. https://example.com:12322/)? 

IIRC currently (as of v13.0 anyway) phpPgAdmin is available on both port 80 & 443 (standard http & https) and 12322 (custom https) as well... But perhaps I'm wrong?

**Comment by l-arnold** [comment]
That would work.  I would recomend 12322 and 12323 actually.  Leave the Web Root Open to something else or have it not be visible.  I don't think you want to invite people to land on your database server interface (closed or not).

**Comment by DocCyblade** [comment]
So whats the use case for these db servers, are they only db servers? Most DBAs will connect with a database tool of some sort. TBH i don't use the phpmysql or phppgadmin. I have a GUI util I connect to it. Since it's a stand alone db server it will be accepting outside connections I could just use a command line tool as well. 

I agree that whatever used should not be on port 80, maybe the standard Turnkey page we have seen before could be on 80 unless that is going away. 

Also, as far as lighty or apache what web server takes the least server resources since this is a standalone db? I am thinking lighty was a smaller footprint than apache. Also If you have apache you just have a lapp server.

**Comment by JedMeister** [comment]
As I just posted, sounds great! :+1:

**Comment by l-arnold** [comment]
When running in production the server is either to be connected to, fromntheboutside,nor to be built upon.  Totally correct about lightly and apache doing nearly the same.  I have seen arguments for using NINE and I believe the tKL Tendenci build uses lightly plus nine for redirects.

Such an approach mirrors the Apache redirect method in my current Odoo.

Anyway, my proposal is:

Port 80 https:// redirect to 443 webconsole
443 https:// webconsole www/index.php
port 21320  https webshell
port 12321 https:// Webin 
port 12322 https:// new standard, Adminer
port 12323 https:// pH PhpPgAdmin
Port 12324 https:// redirect to https:// 12325 Web console
12325  https:// Web console ie www/turnkey/ index.php

Concept here is the whole system cannot be sniffed. Only on https://

Www/index.php can be replaced and console stays in place on 12325.  Could start with  www.www/index.phone being a symbol link to www/turnkey/index.phone

Sym link would be deleted if used for a real index.phone of some other build.

Need to be sure all other services, such as Cron, can work if port 80 redirects.

Ask S
Along and Liraz. 
Unfortunately I don't know how to work with lightly well.

Same distribution should also be in LAPP imho.

Best to all.

**Comment by DocCyblade** [comment]
I don't know lighttpd very well however I'll take this chance to learn it. I am only working on the adminer/lighttpd config for now. As long as I use the same port that phppgadmin used I don't see an issue. 


So to be clear, is this a yes to adminer with lighttpd ?

**Comment by l-arnold** [comment]
Easiest is to use the port that phpgadmin used for Adminer consistency. 

 I will look at it again to see if there are some Config files that can be copied first and reported such that both Admin systems have a port.  (I have some Active Lapp installs I can look at no matter, but will pull an Git for the App right now to be safe)

TBH I need mainly to learn about branches rather than Clones.  All my work to date has been solo in Clones.  Thanks for working on this.

**Comment by l-arnold** [comment]
I seem to have stumbled my way through Making a Fork, a Branch, A branch to my branch (likely unneeded), followed by a Merge to my Branch.

In it, Port 80 in lightly is now 12323

I am not finding any other "listening references" ala Apache so perhaps this is a good start.

**Comment by JedMeister** [comment]
@DocCyblade 
>I don't know lighttpd very well however I'll take this chance to learn it. I am only working on the adminer/lighttpd config for now. As long as I use the same port that phppgadmin used I don't see an issue.

FWIW I don't know it very well either but my work on the Fileserver appliance has forced me to learn! :smile: 

@l-arnold  Sounds like you're learning a bit more about Git and Lighty too. Good work. :+1: 

Both of you (and anybody else who might be reading this) I have added a couple of things that I have learned about Lighty to the TKL docs here: http://www.turnkeylinux.org/docs/lighttpd

Both of you please feel free to add to that page as you see fit. The TurnKey docs are a wiki and you should be able to edit them when logged into the site. Also anything else that you think is interesting and/or worth sharing can be included in the docs. It doesn't need to be TurnKey specific; just so long as it's TurnKey relevant.  Obviously try to keep it concise and to the point though. Also please avoid posting comments there where possible; although if need be you can....  If you have any issues please let me know.

**Comment by DocCyblade** [comment]
Since I was using this app as a test bed for adding lighttpd to common this is pretty much done.

Need to do some Postgres tweaking but the heavy lifting is done

Work pending on https://github.com/turnkeylinux/common/pull/35

**Comment by DocCyblade** [comment]
Started work on Customize adminer install target PostgreSQL. See commits for details

**Comment by DocCyblade** [comment]
Researching on how to extent adminer found this. Maybe able to use this as an include to customize adminer to our liking.

http://www.adminer.org/en/extension/

**Comment by JedMeister** [comment]
Nice one Ken! :+1:

**Comment by JedMeister** [comment]
I just did a test install of PostgreSQL appliance (using your common and postgresql repos) and it looks like it all works sweet. As I posted a little earlier (https://github.com/turnkeylinux/tracker/issues/363#issuecomment-118211693) I think that the longer term plan would be to condense the tklwebcp index.html to something generic then each appliance adjust it or overwrite it (as whatever is easier). But we'll leave that for now...

But I think that we could still recycle the relevant common index.html (i.e. lmap or lapp) for the purposes of the PostgreSQL & MySQL appliances. What do you think @DocCyblade?

**Comment by DocCyblade** [comment]
That sounds good. Lay a common core html file the modify. If we modify the common html right with some key place holders would make sed a bit easier to do. 

Going to try and finish up lamp/lapp tweaks this weekend. That should make Postgres and MySQL apps easy to do as will reuse some of the code. I am shooting to have pull requests submitted by the end of the weekend!

**Comment by DocCyblade** [comment]
Working on this over the next few days/evenings. It looks like I was working towards editing each app alone. Since the changes were made in common (Change default database, localhost) all that needs to be done is a web control panel default. 

@JedMeister - Looking at the history here looks like we were going to use some of the lamp/lapp html file and the tklwebcp overlay (https://github.com/turnkeylinux/tracker/issues/363#issuecomment-118211693) I am going to use the lapp html file as a base and create an overlay we could use in the future and use sed to customize it.

**Comment by DocCyblade** [comment]
Made some progress this evening. I took the idea for the #363 and created a base file https://github.com/turnkeylinux/tracker/issues/363#issuecomment-121777089

Full details are in the commits, see link at the top

However I am trying to append using sed, this is really got me stumped.

Trying to append an html snippet

```php
                    <div>
                        <a href="https://<?php print
                        $_SERVER{'HTTP_HOST'}; ?>:12322"><img
                        src="images/adminer.png"/>Adminer</a>
                    </div>
```

There are a lot of char that need escaping. Is there a best method to appending a chunk of html in a file using sed?

**Comment by JedMeister** [comment]
Not that i am aware of sorry... 

Also (like I wrote or at least hinted somewhere else), now that we are using Adminer, that snippet will be the same for LAMP, LAPP, MySQL & PostrgeSQL appliances... Actually it also applies to Lighttpd & Nginx too. There will be other differences still (app name, references etc) but mostly they're identical...

**Comment by DocCyblade** [comment]
I think I'll just make them static for now. It is the easy thing to do. And with that I should be able to complete Postgres and MySQL tonight.

**Comment by JedMeister** [comment]
Yeah that's what I thought when I did the ones for Lighty & Nginx

**Comment by DocCyblade** [comment]
Doing final testing, should be able to do a pull request tonight

**Comment by JedMeister** [comment]
I left some notes on the PR. Great work mate. Hugely appreciated! :+1:

**Comment by DocCyblade** [comment]
Closing this should be complete

**Comment by xanderdin** [comment]
Hi guys! Any chance to get phpPgAdmin back? My users are upset. Adminer does not provide enough functionality compared to phpPgAdmin. I created a <a href='https://github.com/xanderdin/postgresql'>fork</a> adding it back (listening on port 12323). It isn't hard for me to keep it for my users. But may be you're going to change your mind too? Would you mind of a pull request?

**Comment by JedMeister** [comment]
Hi @xanderdin - How about you [open a new issue](https://github.com/turnkeylinux/tracker/issues/new)?

FYI at this point the v14.0 images are finailsed so it won't be happening anytime real soon. However I'll endeavour to discuss it further with Alon and if it has it's own issue then it won't get forgotten. Also if we were to reintroduce phpPgAdmin I'd be inclined to swap Adminer back out again (rather than have both).

It'd also be useful to know the sort of things that Adminer doesn't support. So then we can make an informed decision with a little clearer idea of how much of a problem this is likely to be and how many users this might be affecting/upsetting.

Also something else that may be worth doing is to contact Adminer (upstream) and ask them to include the feature(s) that you and your users are missing. Even if we do go back to phpPgAdmin it'd still probably be good feedback for them.

**Comment by xanderdin** [comment]
Ok, I got your point. I feel like it doesn't need to be returned if it doesn't affect anyone else. Just notified that in case of 'if' there's someone who has the job already done :). That's why I'm not opening a new issue.

What does Adminer lack? It's easier to open both of them together and compare. Please, try to find user defined functions in a postgres DB, try to view and change triggers, try to restore dumps of hundreds of megabytes, try to list and work with DB roles, try to list and change permissions, compare the number of information presented in both, and so on.

In v13 we used both webmin-postgresql and phpPgAdmin. They complement each other because each of them has its own weaknesses too. I believe Adminer has its strengths and would complement those two. But not as a replacement yet (IMHO).

**Comment by DocCyblade** [comment]
I could be wrong but is there license issue with using one or the other. IIRC I think there was some license reason but don't remember the details. @JedMeister do you remember ?

**Comment by JedMeister** [comment]
@xanderdin - Always appreciate input and feedback! So thanks for taking the time. :smile: And thanks for the roundup of "missing features" in Adminer; I'm not a big Postgres user (actually DBs in general...) so that is very useful info.

As I said is it what it is for v14.0 but we may re-visit this for v14.1. Perhaps we'll wait and see. Maybe we can just document swapping them back for those that want to?

Also FYI we rebuilt the postgres appliance; now including the missing Webmin module... I did it earlier in the week and Alon has signed and uploaded it but it can sometimes take a day or 2 for the mirrors to be totally synced.

@DocCyblade - don't recall licencing issues...  As I recall it; Liraz did a feature request/suggestion to replace both phpMyAdmin & phpPgAdmin with Adminer some time ago. 

What really sparked the change was a problem with phpMyAdmin refusing to stay bound to port 12322 (on appliances that include MySQL). Replacing phpMyAdmin with Adminer was a no-brainer in that context. Then seeing as it also supported Postgres, we figured that replacing phpPgAdmin would also make sense. Then you did the heavy lifting! :smile: 

But maybe the Adminer support for Postgres is not enough? I guess we will wait and see...

**Comment by aaronkempf** [comment]
Wow. From what I can understand.. I think that adminer is a step down on
both the Mysql AND the postgres side.

I think every appliance should have all three Adminer, phpMy AND phppg

I don't see the merits of adminer.  And I'm frustrated that the ports are
inconsistent throughout the apps.

What port do I use for databases on the WordPress app, and on the lapp
appliance ?

As a DBA.  I was always frustrated by phpMy. Now.. Adminer seems even
worse.

**Comment by l-arnold** [comment]
I think this is an example of how "simple extension scriipts" would solve different things on different appliances.  Basically it would be quite easy to have a small package that loads to any of the DB Apps that added additional manager software.

Similar to what we have discussed towards extending some of the CRM/Accounting type apps (Odoo for instance).

The beauty of a "script" is that it should assure it always goes in the same folder structure etc then when migrating to a new build, it couild be reset pretty easily as well.

**Comment by JedMeister** [comment]
@aaronkempf - Thanks for your feedback. As I noted above we had (IMO a good) reason for ditching phpMyAdmin. Beyond your comment (and IIRC one other in the forums) we have not had negative comment regarding the replacement of phpMyAdmin. In fact we have had quite a bit of positive feedback (including a few confirmed phpMyAdmin lovers). However it does seem that Adminer lacks features that are included in phpPgAdmin. 

@l-arnold - Yes you raise a good point. Perhaps we could start a repo of community provided "tweak" scripts? IMO the first step is to document the required steps. Once that's done developing a script is usually pretty easy...
